### PR TITLE
Reuse registers that are cleared.

### PR DESCRIPTION
**Pull request by maierlars**

### CODE REVIEW

File: arangod/Aql/ExecutionBlockImpl.cpp [hunk 1]
```diff
@@ -185,6 +185,9 @@ std::unique_ptr<OutputAqlItemRow> ExecutionBlockImpl<Executor>::createOutputRow(
     for (auto const& reg : *_registerInfos.getOutputRegisters()) {
       for (size_t row = 0; row < newBlock->size(); row++) {
         AqlValue const& val = newBlock->getValueReference(row, reg);
+        if (!val.isEmpty()) {
+          LOG_DEVEL << "register " << reg << " not empty, found: " << val.slice().toJson();
+        }
```
Comment (goedderz) [inline arangod/Aql/ExecutionBlockImpl.cpp:190]:
```suggestion
```

**Comment by maierlars** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-simple-performance--ANY-linux/1128/

**Comment by maierlars** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/9610/

**Comment by maierlars** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/9612/

File: arangod/Aql/RegisterPlan.cpp [hunk 2]
```diff
@@ -199,16 +216,41 @@ void RegisterPlanT<T>::increaseDepth() {
...
+
+template <typename T>
+void RegisterPlanT<T>::registerVariable(VariableId v, std::unordered_set<RegisterId>& unusedRegisters) {
+  RegisterId regId;
+
+  if (unusedRegisters.empty()) {
+    regId = totalNrRegs;
+    addRegister();
+  } else {
+    auto iter = unusedRegisters.begin();
```
Comment (goedderz) [inline arangod/Aql/RegisterPlan.cpp:232]: As this is an `unordered_set`, this line makes the register planning non-deterministic. I'd prefer it to make it deterministic, so maybe use a `set` instead?

File: arangod/Aql/RegisterPlan.cpp [hunk 3]
```diff
@@ -199,16 +216,41 @@ void RegisterPlanT<T>::increaseDepth() {
...
+        std::string("duplicate register assignment for variable #") +
+            std::to_string(v) + " while planning registers");
+  }
 }
 
 template <typename T>
 void RegisterPlanT<T>::registerVariable(VariableId v) {
-  auto total = totalNrRegs;
+  auto regId = totalNrRegs;
   addRegister();
```
Comment (goedderz) [inline arangod/Aql/RegisterPlan.cpp:251]:
```suggestion
  auto regId = addRegister();
```

File: arangod/Aql/RegisterPlan.cpp [hunk 4]
```diff
@@ -199,16 +216,41 @@ void RegisterPlanT<T>::increaseDepth() {
 }
 
 template <typename T>
-void RegisterPlanT<T>::addRegister() {
+RegisterId RegisterPlanT<T>::addRegister() {
```
Comment (goedderz) [inline arangod/Aql/RegisterPlan.cpp:219]: 👍

File: arangod/Aql/ExecutionNode.cpp [hunk 5]
```diff
@@ -1198,48 +1200,34 @@ RegisterId ExecutionNode::variableToRegisterOptionalId(Variable const* var) cons
...
-      }
-    }
-  }
 
+  RegisterId const nrRegs = getRegisterPlan()->nrRegs[getDepth()];
   auto regsToClear = getRegsToClear();
+  auto regsToKeep = calcRegsToKeep();
 
   auto registerInfos = RegisterInfos{make_shared_unordered_set(),
                                      make_shared_unordered_set(),
```
Comment (goedderz) [inline arangod/Aql/ExecutionNode.cpp:1233]: Very good to get rid of this special case!

And at this point you can just call
```
auto registerInfos = createRegisterInfos(make_shared_unordered_set(), make_shared_unordered_set());
```
which calls `calcRegsToKeep` etc.

**Comment by maierlars** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/9617/

**Comment by maierlars** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/9621/

**Comment by maierlars** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/9637/

**Comment by maierlars** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/9647/

**Comment by maierlars** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/9656/

**Comment by maierlars** [comment]
image
image
```
* Test "resilience_move"
    [FAILED]  tests/js/server/resilience/move/moving-shards-cluster.js

      "test" failed: Cleanup missing - test left over collections / views : UnitTestMovingShards33,UnitTestMovingShards34,UnitTestMovingShards35 / . Original test status: {"setUpAllDuration":{"MovingShardsSuite_nodata":0,"MovingShardsSuite_data":0},"testSetup_nodata":{"setUpDuration":325,"tearDownDuration":162,"status":true,"duration":12},"testShrinkNoReplication_n
```

**Pull request merged**